tfhe/src/integer/server_key/radix_parallel/slice.rs

```diff
@@ -0,0 +1,371 @@
+    ///
+    /// The result is returned as a new ciphertext. This function is more efficient
+    /// if the range starts on a block boundary.
+    ///
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+    ///
+    /// // We have 4 * 2 = 8 bits of message
+    /// let num_blocks = 4;
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2_KS_PBS, num_blocks);
+    ///
+    /// let msg: u64 = 225;
+    /// let start_bit = 3;
+    /// let end_bit = 6;
+    ///
+    /// // Encrypt the message:
+    /// let ct = cks.encrypt(msg);
+    ///
+    /// let ct_res = sks
+    ///     .unchecked_scalar_bitslice_parallelized(&ct, start_bit..end_bit)
+    ///     .unwrap();
+    ///
+    /// // Decrypt:
+    /// let clear = cks.decrypt(&ct_res);
+    /// assert_eq!((msg % (1 << end_bit)) >> start_bit, clear);
+    /// ```
+    pub fn unchecked_scalar_bitslice_parallelized<B, R>(
+        &self,
+        ctxt: &RadixCiphertext,
+        range: R,
+    ) -> Result<RadixCiphertext, InvalidRangeError>
+    where
+        R: RangeBounds<B>,
+        B: CastFrom<usize> + CastInto<usize> + Copy,
+    {
+        let block_width = self.message_modulus().0.ilog2() as usize;
+        let range = normalize_range(&range, block_width * ctxt.blocks.len())?;
+
+        let slice_width = range.end - range.start;
+
+        // If the starting bit is block aligned, we can do most of the slicing with block copies.
+        // If it's not we must extract the bits with PBS. In either cases, we must extract the last
+        // bits with a PBS if the slice size is not a multiple of the block size.
+        let mut sliced = if range.start % block_width != 0 {
+            let mut sliced = self.scalar_blockslice_unaligned_parallelized(
+                ctxt,
+                range.start / block_width,
+                slice_width / block_width,
+                range.start % block_width,
+            );
+
+            if slice_width % block_width != 0 {
+                let last_block = self.bitslice_remainder_unaligned(
+                    ctxt,
+                    range.start / block_width + slice_width / block_width,
+                    range.start % block_width,
+                    slice_width % block_width,
+                );
+                sliced.blocks.push(last_block);
+            }
+
+            sliced
+        } else {
+            let mut sliced = self.scalar_blockslice_aligned(
+                ctxt,
+                range.start / block_width,
+                range.end / block_width,
+            );
+            if slice_width % block_width != 0 {
+                let last_block = self.bitslice_remainder(
+                    ctxt,
+                    range.end / block_width,
+                    slice_width % block_width,
+                );
+                sliced.blocks.push(last_block);
+            }
+
+            sliced
```

Comment:
The computations inside the `if` should be done in a rayon::join/rayon::scope, otherwise, when `slice_width % block_width != 0`, the latency is 2 PBS, with rayon::join/rayon::scope we could keep the latency at 1 PBS

Reply:
done